sandbox/apparmor/notify/listener/listener.go

```diff
@@ -0,0 +1,365 @@
+// Package listener implements a high-level interface to the apparmor
+// notification mechanism. It can be used to build userspace applications
+// which respond to apparmor prompting profiles.
+package listener
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"sync/atomic"
+
+	"gopkg.in/tomb.v2"
+
+	"github.com/snapcore/snapd/logger"
+	"github.com/snapcore/snapd/osutil/epoll"
+	"github.com/snapcore/snapd/sandbox/apparmor/notify"
+)
+
+var (
+	// ErrListenerAlreadyRun indicates that the Run() method has already been
+	// run. Each listener must only be run once, so that spawned goroutines can
+	// be safely tracked and terminated.
+	ErrListenerAlreadyRun = errors.New("listener has already been run")
+
+	// ErrListenerAlreadyClosed indicates that the listener has been closed
+	ErrListenerAlreadyClosed = errors.New("listener has already been closed")
+
+	// ErrListenerNotSupported indicates that the kernel does not support apparmor prompting
+	ErrListenerNotSupported = errors.New("kernel does not support apparmor notifications")
+
+	osOpen      = os.Open
+	notifyIoctl = notify.Ioctl
+)
+
+// Request is a high-level representation of an apparmor prompting message.
+//
+// Each request must be replied to by writing a boolean to the YesNo channel.
+type Request struct {
+	// Pid is the identifier of the process triggering the request.
+	Pid uint32
+	// Label is the apparmor label on the process triggering the request.
+	Label string
+	// SubjectUID is the UID of the subject that triggered the prompt
+	SubjectUid uint32
+
+	// Path is the path of the file, as seen by the process triggering the request.
+	Path string
+	// Permission is the opaque permission that is being requested.
+	Permission interface{}
```

Comment:
should all these value be encapsulated to make them read-only? or maybe the doc comment should make it clear that the consumer shouldn't mutate the struct?

Reply:
I like the idea of encapsulating them. Also, some will likely change when we abstract the permissions and/or add more mediation classes (i.e. things in addition to just file operations). Methods will probably be useful for that anyway so we can do error handling and permission wrangling, so I'll go ahead and encapsulate all of these (or perhaps make some private if they're not used?)

Reply:
`Pid` isn't used (yet?), but I added a function to export it anyway. And renamed it to `PID()` to align with go style. Similarly, `s/SubjectUid/SubjectUID/g`.